lib/compat.php

```diff
@@ -77,6 +77,84 @@ function gutenberg_provide_render_callback_with_block_object( $pre_render, $bloc
 }
 add_filter( 'pre_render_block', 'gutenberg_provide_render_callback_with_block_object', 10, 2 );
 
+/**
+ * Registers the user meta property responsible for storing preferences for the
+ * data script persistence middleware. The meta is explicitly registered so that
+ * it is made accessible for read and write via the REST API.
+ */
+function gutenberg_register_data_persistence_user_meta() {
+	global $wpdb;
+	register_meta(
+		'user',
+		$wpdb->prefix . 'data_persistence',
+		array(
+			'type'         => 'object',
+			'single'       => true,
+			'show_in_rest' => array(
+				'name'   => 'data_persistence',
+				'type'   => 'object',
+				'schema' => array(
+					'type'                 => 'object',
+					'properties'           => array(),
+					'additionalProperties' => true,
+				),
+			),
+		)
+	);
+}
+add_action( 'init', 'gutenberg_register_data_persistence_user_meta' );
+
+/**
+ * Enqueues inline scripts used to provide a custom storage interface for the
+ * data script persistence middleware. The custom storage implementation stores
+ * persisted values using user meta, bootstrapped for initial read at load-time,
+ * and updated via the REST API on change.
+ */
+function gutenberg_user_settings_data_persistence_inline_script() {
+	global $wp_scripts, $wpdb;
+
+	$user_id         = get_current_user_id();
+	$persisted_value = get_user_meta( $user_id, $wpdb->prefix . 'data_persistence', true );
+	if ( empty( $persisted_value ) ) {
+		// If there's no explicit metadata assigned, fall back to a value which
+		// was persisted using browser storage, prior to user meta persistence.
+		$persisted_value = sprintf( 'localStorage.getItem( "WP_DATA_USER_%s" );', $user_id );
+	} else {
+		// Otherwise, encode the string value for interpolation in the storage
+		// implementation script. The first `json_encode` will is responsible
+		// for producing a JSON encoding of the persisted meta object, and the
+		// second will apply quoting to that string result.
+		$persisted_value = json_encode( json_encode( $persisted_value ) );
```

Comment:
```suggestion
		$persisted_value = wp_json_encode( wp_json_encode( $persisted_value ) );
```
Coding standards says we should use `wp_json_encode` instead of `json_encode`